content/posts/optree/pytrees/index.md

```diff
@@ -0,0 +1,203 @@
+print(f"{all_positive=}")
+# >> all_positive=True
+
+summed = pt.tree_reduce(sum, pytree)
+print(f"{summed=}")
+# >> summed=array([13.])
+```
+
+The trick here is that these operations can be implemented in three steps, e.g. `tree_map`:
+
+```python
+# step 1:
+leafs, treedef = pt.tree_flatten(pytree)
+
+# step 2:
+new_leafs = tuple(map(fun, leafs))
+
+# step 3:
+result_pytree = pt.tree_unflatten(treedef, new_leafs)
+```
+
+Here, we use [`optree`](https://github.com/metaopt/optree/tree/main/optree) - a standalone PyTree library - that enables all these manipulations. It focusses on performance, feature richness, minimal dependencies, and got adopted by [PyTorch](https://pytorch.org), [Keras](https://keras.io), and [TensorFlow](https://github.com/tensorflow/tensorflow) (through Keras) as a core dependency.
+
+### PyTree Origins
+
+Originally, the concept of PyTrees was developed by the [JAX](https://docs.jax.dev/en/latest/) project to make nested collections of JAX arrays work transparently at the "JIT-boundary" (the JAX JIT toolchain does not know about Python containers, only about JAX Arrays).
+However, PyTrees were quickly adopted by AI researchers for broader use-cases: semantically grouping layers of weights and biases in e.g. a list of named tuples (or dictionaries) is a common pattern in the JAX-AI-world, see the following (pseudo) Python snippet:
+
+```python
+from typing import NamedTuple, Callable
+import jax
+import jax.numpy as jnp
+
+
+class Layer(NamedTuple):
+    W: jax.Array
+    b: jax.Array
+
+
+layers = [
+    Layer(W=jnp.array(...), b=jnp.array(...)),  # first layer
+    Layer(W=jnp.array(...), b=jnp.array(...)),  # second layer
+    ...,
+]
+
+
+@jax.jit
+def neural_network(layers: list[Layer], x: jax.Array) -> jax.Array:
+    for layer in layers:
+        x = jnp.tanh(layer.W @ x + layer.b)
+    return x
+
+
+prediction = neural_network(layers=layers, x=jnp.array(...))
+```
+
+Here, `layers` is a PyTree - a `list` of multiple `Layer` - and the JIT compiled `neural_network` function _just works_ with this datastructure as input.
+
+### PyTrees in Scientific Python
+
+Wouldn't it be nice to make workflows in the scientific Python ecosystem _just work_ with any PyTree?
+
+Giving semantic meaning to numeric data through PyTrees can be useful for applications outside of AI as well.
+Consider the following minimization of the [Rosenbrock](https://en.wikipedia.org/wiki/Rosenbrock_function) function:
+
+```python
+from scipy.optimize import minimize
+
+
+def rosenbrock(params: tuple[float]) -> float:
+    """
+    Rosenbrock function. Minimum: f(1, 1) = 0.
+
+    https://en.wikipedia.org/wiki/Rosenbrock_function
+    """
+    x, y = params
+    return (1 - x) ** 2 + 100 * (y - x**2) ** 2
+
+
+x0 = (0.9, 1.2)
+res = minimize(rosenbrock, x0)
+print(res.x)
+# >> [0.99999569 0.99999137]
+```
+
+Now, let's turn it a minimization that uses a more complex type for the parameters - a NamedTuple that describes our fit parameters:
```

Comment:
```suggestion
Now, let's consider a minimization that uses a more complex type for the parameters &mdash; a NamedTuple that describes our fit parameters:
```

I'm not sure what the best rephasing is here for an alternative to "turn it". Maybe "let's consider" or "let's use" or "let's see"? Maybe I'm missing the intent of the idea you have.